src/lib/xmlparse.rs

```diff
@@ -2821,101 +2803,79 @@ pub unsafe extern "C" fn XML_ParseBuffer(
+                    self.m_bufferStart -= offset;
                 }
             } else {
-                let mut bufferSize: c_int = match safe_ptr_diff(self.m_bufferLim, self.m_bufferPtr) {
+                let mut bufferSize: c_int = match self.m_buffer.len() - self.m_bufferStart {
                     0 => INIT_BUFFER_SIZE,
                     size => size.try_into().unwrap(),
                 };
-                while bufferSize < neededSize {
+                while (bufferSize as usize) < neededSize {
                     bufferSize = match 2i32.checked_mul(bufferSize) {
                         Some(s) => s,
                         None => {
                             self.m_errorCode = XML_ERROR_NO_MEMORY;
-                            return ptr::null_mut();
+                            return None;
                         }
                     }
                 }
-                let newBuf = MALLOC![c_char; bufferSize];
-                if newBuf.is_null() {
+                if self.m_buffer.try_reserve(bufferSize as usize).is_err() {
                     self.m_errorCode = XML_ERROR_NO_MEMORY;
-                    return ptr::null_mut();
+                    return None;
                 }
-                self.m_bufferLim = newBuf.offset(bufferSize as isize);
-                if !self.m_bufferPtr.is_null() {
-                    memcpy(
-                        newBuf as *mut c_void,
-                        &*self.m_bufferPtr.offset(-keep as isize) as *const c_char
-                            as *const c_void,
-                        (safe_ptr_diff(self.m_bufferEnd, self.m_bufferPtr) as c_long + keep as c_long) as c_ulong,
-                    );
-                    FREE!(self.m_buffer);
-                    self.m_buffer = newBuf;
-                    self.m_bufferEnd = self
-                        .m_buffer
-                        .offset(safe_ptr_diff(self.m_bufferEnd, self.m_bufferPtr))
-                        .offset(keep as isize);
-                    self.m_bufferPtr = self.m_buffer.offset(keep as isize)
+                self.m_buffer.resize(bufferSize as usize, 0);
+                if self.m_bufferStart < self.m_bufferEnd {
+                    self.m_buffer.copy_within(self.m_bufferStart-keep..self.m_bufferEnd, 0);
+                    self.m_bufferEnd = self.m_bufferEnd - self.m_bufferStart + keep;
+                    self.m_bufferStart = keep;
                 } else {
                     /* This must be a brand new buffer with no data in it yet */
-                    self.m_bufferEnd = newBuf;
-                    self.m_buffer = newBuf;
-                    self.m_bufferPtr = self.m_buffer
+                    self.m_bufferStart = 0;
+                    self.m_bufferEnd = 0;
                 }
             }
             self.m_eventEndPtr = ptr::null();
             self.m_eventPtr = ptr::null();
-            self.m_positionPtr = ptr::null();
+            self.m_positionIdx = 0;
```

Comment:
I wonder if `m_positionIdx` should have the `Option<usize>` type to distinguish between "no index" and "index 0".

Reply:
I thought about this, yeah, that might be good. It makes some other pointer operations harder, though.

What I kind of what to do is put all the indices into a struct to encapsulate them along with the buffer itself so that we're guaranteed to correctly update them when we resize and copy inside the buffer.